python/dgl/distributed/rpc.py

```diff
@@ -0,0 +1,487 @@
+
+    @property
+    def msg_seq(self):
+        """Get message sequence number."""
+        return _CAPI_DGLRPCMessageGetMsgSeq(self)
+
+    @property
+    def client_id(self):
+        """Get client ID."""
+        return _CAPI_DGLRPCMessageGetClientId(self)
+
+    @property
+    def server_id(self):
+        """Get server ID."""
+        return _CAPI_DGLRPCMessageGetServerId(self)
+
+    @property
+    def xxx(self):
+        return 1
+
+    @property
+    def data(self):
+        """Get payload buffer."""
+        return _CAPI_DGLRPCMessageGetData(self)
+
+    @property
+    def tensors(self):
+        """Get tensor payloads."""
+        rst = _CAPI_DGLRPCMessageGetTensors(self)
+        return [F.zerocopy_from_dgl_ndarray(tsor.data) for tsor in rst]
+
+def send_request(target, request):
+    """Send one request to the target server.
+
+    Serialize the given request object to an :class:`RPCMessage` and send it
+    out.
+
+    The operation is non-blocking -- it does not guarantee the payloads have
+    reached the target or even have left the sender process. However,
+    all the payloads (i.e., data and arrays) can be safely freed after this
+    function returns.
+
+    Parameters
+    ----------
+    target : int
+        The server ID.
+    request : Request
+        The request to send.
+
+    Raises
+    ------
+    ConnectionError if there is any problem with the connection.
+    """
+    service_id = request.service_id
+    msg_seq = incr_msg_seq()
+    client_id = get_rank()
+    server_id = target
+    data, tensors = serialize_to_payload(request)
+    msg = RPCMessage(service_id, msg_seq, client_id, server_id, data, tensors)
+    send_rpc_message(msg)
+
+def recv_response(timeout=0):
+    """Receive one response.
+
+    Receive one :class:`RPCMessage` and de-serialize it into a proper Response object.
+
+    The operation is blocking -- it returns when it receives any message
+    or it times out.
+
+    Parameters
+    ----------
+    timeout : int, optional
+        The timeout value in milliseconds. If zero, wait indefinitely.
+
+    Returns
+    -------
+    res : Response
+        One response received from the target, or None if it times out.
+
+    Raises
+    ------
+    ConnectionError if there is any problem with the connection.
+    """
+    msg = recv_rpc_message(timeout)
+    if msg is None:
+        return None
+    _, res_cls = SERVICE_ID_TO_PROPERTY[msg.service_id]
+    if res_cls is None:
+        raise DGLError('Got response message from service ID {}, '
+                       'but no response class is registered.'.format(msg.service_id))
+    res = deserialize_from_payload(res_cls, msg.data, msg.tensors)
+    if res.client_id != get_rank():
+        raise DGLError('Got reponse of request sent by client {}, '
+                       'different from my rank {}!'.format(res.client_id, get_rank()))
+    return res
+
+def remote_call(target_and_requests, timeout=0):
```

Comment:
if we want to support async pull, we can still support it with the current network API, right?

Reply:
It will probably be another API called `remote_call_async` and it will return a list of future objects to get the responses asynchronously.